compiler/passes/convert-typed-uast.cpp

```diff
@@ -0,0 +1,2041 @@
+/*
+ * Copyright 2020-2024 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* This file supports conversion of uAST+types to the older AST used by
+   the rest of the compiler. It uses recursive functions to perform
+   the conversion.
+
+   This pass is currently experimental. While it is optional, it should
+   generate AST similar to what the production compiler emits after
+   the callDestructors pass.
+ */
+
+#include "convert-uast.h"
+
+#include "CForLoop.h"
+#include "CatchStmt.h"
+#include "DecoratedClassType.h"
+#include "DeferStmt.h"
+#include "DoWhileStmt.h"
+#include "ForLoop.h"
+#include "ForallStmt.h"
+#include "IfExpr.h"
+#include "ImportStmt.h"
+#include "LoopExpr.h"
+#include "ParamForLoop.h"
+#include "ResolveScope.h"
+#include "TemporaryConversionThunk.h"
+#include "TryStmt.h"
+#include "WhileDoStmt.h"
+#include "build.h"
+#include "config.h"
+#include "global-ast-vecs.h"
+#include "metadata.h"
+#include "optimizations.h"
+#include "parser.h"
+//#include "passes.h"
+#include "resolution.h"
+#include "stmt.h"
+
+
+#include "view.h" // DEBUG only
+
+#include "chpl/framework/compiler-configuration.h"
+#include "chpl/framework/global-strings.h"
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/ResolvedVisitor.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/types/all-types.h"
+#include "chpl/uast/all-uast.h"
+#include "chpl/uast/chpl-syntax-printer.h"
+#include "chpl/util/assertions.h"
+#include "chpl/util/string-escapes.h"
+
+#include "convert-help.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
+
+#include <iostream>
+
+using namespace chpl;
+using namespace resolution;
+using namespace uast;
+
+// converts resolved uAST + types to typed AST
+struct TConverter final : UastConverter {
+
+  /// Nested Types ///
+  using RV = ResolvedVisitor<TConverter>;
+
+  // State required to mimic the desugaring of multi-declarations into
+  // regular ones.
+  struct MultiDeclState {
+    // The result of computing the target locale, stored in a temp, used
```

Comment:
This is duplicated verbatim from `convert-uast.cpp`. Consider making it an associated / inner class of `UastConverter` or just lifting it to a top-level struct.

Reply:
I will add a TODO comment. The issue is that the typed converter doesn't yet handle variable multi-decls, and while it probably needs something kindof like this, it's very unlikely to be exactly this. In particular, it should be using some sort of `Type*` / `QualifiedType` rather than `Expr* prevTypeExpr`.